tests/foreman/ui_airgun/test_host.py

```diff
@@ -134,6 +145,248 @@ def module_host_template(module_org, module_loc):
+        'Kickstart default iPXE',
+        'Kickstart default',
+        'Kickstart default finish',
+        'Kickstart default user data'
+    ]:
+        template = entities.ConfigTemplate().search(
+            query={
+                u'search': u'name="{}"'.format(template_name)
+            }
+        )[0].read()
+        template.operatingsystem.append(os)
+        template.organization.append(module_org)
+        template.location.append(module_loc)
+        template = template.update([
+            'location', 'operatingsystem', 'organization'])
+        templates.append(template)
+    # Update the OS to associate architecture, ptable, templates
+    os.architecture = [default_architecture]
+    os.ptable = [default_partition_table]
+    os.config_template = templates
+    os = os.update(['architecture', 'config_template', 'ptable'])
+    return os
+
+
+@pytest.fixture(scope='module')
+def os_path(module_libvirt_os):
+    # Check what OS was found to use correct media
+    if module_libvirt_os.major == str(RHEL_6_MAJOR_VERSION):
+        os_distr_url = settings.rhel6_os
+    elif module_libvirt_os.major == str(RHEL_7_MAJOR_VERSION):
+        os_distr_url = settings.rhel7_os
+    else:
+        raise ValueError('Proposed RHEL version is not supported')
+    return os_distr_url
+
+
+@pytest.fixture(scope='module')
+def module_proxy(module_org, module_loc):
+    # Search for SmartProxy, and associate organization/location
+    proxy = entities.SmartProxy().search(
+        query={u'search': u'name={0}'.format(settings.server.hostname)})[0].read()
+    proxy.location.append(module_loc)
+    proxy.organization.append(module_org)
+    proxy = proxy.update(['location', 'organization'])
+    return proxy
+
+
+@pytest.fixture(scope='module')
+def module_libvirt_resource(module_org, module_loc):
+    # Search if Libvirt compute-resource already exists
+    # If so, just update its relevant fields otherwise,
+    # Create new compute-resource with 'libvirt' provider.
+    resource_url = u'qemu+ssh://root@{0}/system'.format(
+        settings.compute_resources.libvirt_hostname
+    )
+    comp_res = [
+        res for res in entities.LibvirtComputeResource().search()
+        if res.provider == FOREMAN_PROVIDERS['libvirt'] and res.url == resource_url
+    ]
+    if len(comp_res) > 0:
+        computeresource = entities.LibvirtComputeResource(id=comp_res[0].id).read()
+        computeresource.location.append(module_loc)
+        computeresource.organization.append(module_org)
+        computeresource = computeresource.update(['location', 'organization'])
+    else:
+        # Create Libvirt compute-resource
+        computeresource = entities.LibvirtComputeResource(
+            provider=FOREMAN_PROVIDERS['libvirt'],
+            url=resource_url,
+            set_console_password=False,
+            display_type=u'VNC',
+            location=[module_loc],
+            organization=[module_org],
+        ).create()
```

Comment:
Do you think we have to validate that compute resource created is valid here?

Reply:
Nope, as:
1) All these preconditions are very strict. Even if one element goes wrong, the whole scenario will fail. For example:
    a) if compute resource is not valid, host 'Virtual Machine' tab will produce an exception and corresponding field will not be accessible. As result, host will not be created and we catch the problem
    b) if your subnet is not valid, your DHCP server will not provide ip from the pool and you get an exception for your network interface.
    c) if your domain is not valid, you will also get an exception as subnet will not be accessible
2) Such validations were not present in the initial scenario
3) It is API functionality which is always out of scope for WebUI validation


Basically, it makes sense to improve something from WebUI side, like return exact exception rather than fact that we cannot create the host, but it is low priority as we should see the error on a screenshot

Reply:
As all of you module fixtures are part of test settings any failure is not recorded immediately and wait for the next fixture to show if that is failed. While analysing test, it will be each fixture depend on others to show the result. Probably that will make the user search in the wrong side of failure as the test data is invalid for my next fixture.

I see you point 2 & 3 valid enough. But consider a case where All Test setting got passed as there is no validation or assert, UI test will start and ultimately get failed ( This is time consuming based on UI test ). That was my whole point.

Reply:
Once more, how do you propose to validate? `create()` already doing `read()`, so if something goes wrong with create, you will get an exception anyway. But to understand that entity is really valid, assert doesn't suffice our needs and we actually can validate that fact on host create procedure only

Reply:
OK Valid enough.